Cython/Utility/UFuncs_C.c

```diff
@@ -14,6 +14,58 @@
 #define __PYX_PYUFUNCGENERICFUNCTION_CAST(x) (PyUFuncGenericFunction)x
 #endif
 
+/////////////////////// UFuncTypeHandling.proto //////////////
+
+#define __PYX_GET_NPY_COMPLEX_TYPE(tp) \
+    sizeof(tp) == sizeof(npy_cfloat) ? NPY_CFLOAT : \
+    sizeof(tp) == sizeof(npy_cdouble) ? NPY_CDOUBLE : \
+    sizeof(tp) == sizeof(npy_clongdouble) ? NPY_CLONGDOUBLE : NPY_NOTYPE
+
+#define __PYX_GET_NPY_FLOAT_TYPE(tp) \
+    sizeof(tp) == sizeof(npy_float) ? NPY_FLOAT : \
+    sizeof(tp) == sizeof(npy_double) ? NPY_DOUBLE : \
+    sizeof(tp) == sizeof(npy_longdouble) ? NPY_LONGDOUBLE : NPY_NOTYPE
+
+#define __PYX_GET_NPY_UINT_TYPE(tp) \
+    sizeof(tp) == 1 ? NPY_UINT8 : \
+    sizeof(tp) == 2 ? NPY_UINT16 : \
+    sizeof(tp) == 4 ? NPY_UINT32 : \
+    sizeof(tp) == 8 ? NPY_UINT64 : NPY_NOTYPE
+
+#define __PYX_GET_NPY_SINT_TYPE(tp) \
+    sizeof(tp) == 1 ? NPY_INT8 : \
+    sizeof(tp) == 2 ? NPY_INT16 : \
+    sizeof(tp) == 4 ? NPY_INT32 : \
+    sizeof(tp) == 8 ? NPY_INT64 : NPY_NOTYPE
+
+#define __PYX_GET_NPY_INT_TYPE(tp) \
+    (((tp)-1) < (tp)0) ? \
+        (__PYX_GET_NPY_SINT_TYPE(tp)) : \
+        (__PYX_GET_NPY_UINT_TYPE(tp))
+
+static int __Pyx_validate_ufunc_types(char *types, Py_ssize_t count, Py_ssize_t input_count);
+
+/////////////////////// UFuncTypeHandling ///////////////
+
+// DW - it's a bit of a shame that this has to be a runtime check since the C compiler does
+// know it. We could make this a compile-time error in C++, but not easily in C.
```

Comment:
It could potentially figure it out if it inlines the validation function. But how many types do you expect here? One, two, three? Probably rarely more, right? Why not call the validation function multiple times, instead of using a loop? You could also inline the `NPY_NOTYPE` comparison and keep the helper function only for setting the exception. That should make it easy for the C compiler to discard the calls in the normal cases.

Reply:
Ah I think you're slightly misunderstanding the comment.

What I really want is a failure from the C compiler because it already knows that it's used `NPY_NOTYPE`. I'm not worried about the cost of the validation function. I wish I didn't have to try to import the module to work out that I've got it wrong.

Reply:
Yeah possibly. I'll have another look and see if I can do something like that.